### DISCUSSION TITLE
How to condictionly execute an action?

### DISCUSSION THREAD

**Discussion by Jerry-Wang-Developer** [pointfreeco/swift-composable-architecture#1918] [Q&A]
Look forward for best practice for below scenario.

Let say I have a Root state holding login information and some child state.
For child state, some action need to perform after login. that is, if the user is not login, we need to show the login page for user to login, and after login.

The below is a very simple codes.

The code execute order is ③ ① ②  ,  But for my scenario, I want  ③ ① execute and ② is ignore when user is not login, How can I do that?
If can't, What is the best practice to handle this scenario. (Think about that will be many child states and in there lots of action need user login first to forward process.)

And one more question, Is there any way I can access the parent state in child state(If can, move the logic into child state is suit for me too.)

`

struct Root: ReducerProtocol {
    struct State: Equatable, Codable {
        var child: Child.State
        var isLogin: Bool
        var showLoginPage: Bool
    }

    enum Action: Equatable {
        case child(Child.Action)
    }
    
    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .child(.favorite): // ①
                if !state.isLogin {
                    state.showLoginPage = true
                    return .none  
                }
                return .none

            default:
                return .none
            }
        }

        Scope(state: \.child, action: /Action.child) {
            Child()
        }
    }
}

struct Child: ReducerProtocol {
    struct State: Equatable {
        var isFavorite: Bool
    }
    
    enum Action {
        case favorite(Bool)
    }
    
    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case let .favorite(isFavorite): // ②
                state.isFavorite = isFavorite
                return .none
            }
        }
    }
}

struct ChildView: View {
    let store: StoreOf<Child>
    
    var body: some View {
        WithViewStore(store) { viewStore in
            Button {
                viewStore.send(.favorite) // ③
            } label: {
                Text("favorite")
            }
        }
    }
}
`

**Comment by mbrandonw** [comment]
It looks like this is a duplication of #1919 so I am going to lock this one.
